src/shared/watchdog.c

```diff
@@ -308,16 +307,26 @@ static int open_watchdog(void) {
          * has the benefit that we can easily find the matching directory in sysfs from it, as the relevant
          * sysfs attributes can only be found via /sys/dev/char/<major>:<minor> if the new-style device
          * major/minor is used, not the old-style. */
-        fn = !watchdog_device || path_equal(watchdog_device, "/dev/watchdog") ?
-                "/dev/watchdog0" : watchdog_device;
-
-        r = free_and_strdup(&watchdog_device, fn);
-        if (r < 0)
-                return log_oom_debug();
+        if (!watchdog_device || path_equal(watchdog_device, "/dev/watchdog")) {
+                watchdog_fd = open("/dev/watchdog0", O_WRONLY|O_CLOEXEC);
+                if (watchdog_fd < 0) {
+                        bool ignore = watchdog_device;
+
+                        r = log_debug_errno(errno, "Failed to open watchdog device /dev/watchdog0%s: %m", ignore ? ", ignoring" : "");
+                        if (!ignore)
+                                return r;
+                } else {
+                        r = free_and_strdup(&watchdog_device, "/dev/watchdog0");
+                        if (r < 0)
+                                return log_oom_debug();
+                }
+        }
 
-        watchdog_fd = open(watchdog_device, O_WRONLY|O_CLOEXEC);
-        if (watchdog_fd < 0)
-                return log_debug_errno(errno, "Failed to open watchdog device %s, ignoring: %m", watchdog_device);
+        if (watchdog_fd < 0) {
+                watchdog_fd = open(watchdog_device, O_WRONLY|O_CLOEXEC);
+                if (watchdog_fd < 0)
+                        return log_debug_errno(errno, "Failed to open watchdog device %s: %m", watchdog_device);
+        }
```

Comment:
what about this:

```c
char **try_order;

if (!watchdog_device || PATH_IN_SET(watchdog_device, "/dev/watchdog0", "/dev/watchdog"))
        try_order = STRV_MAKE("/dev/watchdog0", "/dev/watchdog");
else
        try_order = STRV_MAKE(watchdog_device);

STRV_FOREACH(wd, try_order) {
        watchdog_fd = open(wd, O_WRONLY|O_CLOEXEC);
        if (watchdog >= 0) {
                r = free_and_strdup(&watchdog_device, wd);
                if (r < 0)
                        return r;
                break; 
        } 
        
        if (errno != ENOENT)
                return log_debug_errno(errno, "Failed to open watchdog device %s: %m", wd);
}

if (watchdog_fd < 0)
       return log_debug_errno(SYNTHETIC_ERRNO(ENOENT), "Failed to open watchdog device %s: %m", watchdog_device ?: "auto");
```
or so?

i.e. always prefer the new-style devices over the old ones, regardless what precisely you specify for the first device.

Reply:
I cannot find big different between the current and suggested code. and unfortunately `STRV_MAKE` in if block cannot used outside.